File: tests/fixtures/generate.py
# Fixture generator, Python half. Run it through `tests/fixtures/generate.mjs`,
# which knows which interpreter each mode needs and why.
#
#   python generate.py parquet <outdir>          needs pyarrow (a venv; see generate.mjs)
#   python generate.py protected-xlsx <outdir>   needs the system python that carries `uno`
#
# Neither tool is a project dependency. These files are generated once, committed,
# and read by the tests as bytes — the shapes below are exactly the ones
# `hyparquet-writer` and `write-excel-file` cannot produce, which is why real
# files exist for them at all.

import datetime
import decimal
import os
import shutil
import subprocess
import sys
import tempfile
import time


def parquet(out):
    """Three Parquet shapes hyparquet-writer cannot write. Produced with pyarrow 25.0.0."""
    import pyarrow as pa
    import pyarrow.parquet as pq

    # 1. INT96 — the legacy timestamp Impala and older Spark still emit. pyarrow
    #    writes it only behind this deprecated flag; nothing in JS writes it.
    pq.write_table(
        pa.table(
            {
                "Erfasst": pa.array(
                    [
                        datetime.datetime(2025, 8, 1, 14, 30, 0),
                        None,
                        datetime.datetime(2024, 2, 29, 0, 0, 0),
                    ],
                    pa.timestamp("us"),
                ),
                "Kunde": pa.array(["Anna", "Bernd", "Clara"]),
            }
        ),
        os.path.join(out, "parquet-int96-timestamp.parquet"),
        use_deprecated_int96_timestamps=True,
        compression="snappy",
    )

    # 2. DECIMAL backed by FIXED_LEN_BYTE_ARRAY. The backing is chosen by
    #    precision: <= 9 gives INT32, <= 18 INT64, and beyond that a byte array.
    #    hyparquet-writer's DECIMAL path accepts only the two integer backings.
    pq.write_table(
        pa.table(
            {
                "Preis": pa.array(
                    [decimal.Decimal("1234567.89"), decimal.Decimal("-0.05"), None],
                    pa.decimal128(38, 2),
                )
            }
        ),
        os.path.join(out, "parquet-decimal-fixed-len-byte-array.parquet"),
        compression="snappy",
    )

    # 3. Two top-level columns of one name, with *different* physical types so a
    #    reader that mixes them up is caught by the values rather than only by a
    #    count. hyparquet-writer resolves columnData against the schema by name
    #    and cannot emit this at all.
    dup_schema = pa.schema([pa.field("Betrag", pa.int64()), pa.field("Betrag", pa.string())])
    pq.write_table(
        pa.Table.from_arrays(
            [pa.array([1, 2]), pa.array(["brutto", "netto"])], schema=dup_schema
        ),
        os.path.join(out, "parquet-duplicate-column-name.parquet"),
        compression="snappy",
    )


def protected_xlsx(out):
    """A genuinely password-protected workbook (password: geheim).

    LibreOffice's `--convert-to` filter options do **not** carry a password for
    xlsx — measured, every documented spelling produced a plain `PK\\x03\\x04`
    zip. The supported route is the UNO API, which is what this does: it starts
    soffice as a listener, loads a CSV and stores it with `Password` and
    `EncryptFile` in the media descriptor.

    The result is an OOXML-encrypted file, which is a CFB container beginning
    `D0 CF 11 E0` rather than a zip — that is what makes it a fixture worth
    having, since the whole question is what a reader does with a workbook it
    cannot open rather than one it can.
    """
    import uno
    from com.sun.star.beans import PropertyValue

    # Scratch space outside the fixture directory: nothing but the finished
    # workbook belongs next to the committed bytes.
    work = tempfile.mkdtemp(prefix="querbeet-fixture-")
    csv_path = os.path.join(work, "quelle.csv")
    with open(csv_path, "w", encoding="utf-8") as handle:
        handle.write("Kunde,Betrag\nAnna,1234.5\nBernd,80\n")

    profile = "file://" + os.path.join(work, "profile")
    soffice = subprocess.Popen(
        [
            "soffice",
            "--headless",
            "--norestore",
            "-env:UserInstallation=" + profile,
            "--accept=socket,host=127.0.0.1,port=2002;urp;",
        ]
    )
    try:
        local = uno.getComponentContext()
        resolver = local.ServiceManager.createInstanceWithContext(
            "com.sun.star.bridge.UnoUrlResolver", local
        )
        ctx = None
        for _ in range(60):
            try:
                ctx = resolver.resolve(
                    "uno:socket,host=127.0.0.1,port=2002;urp;StarOffice.ComponentContext"
                )
                break
            except Exception:
                time.sleep(1)
        if ctx is None:
            raise RuntimeError("no UNO bridge after 60s — is soffice installed?")

        def pv(name, value):
            prop = PropertyValue()
            prop.Name = name
            prop.Value = value
            return prop

        desktop = ctx.ServiceManager.createInstanceWithContext(
            "com.sun.star.frame.Desktop", ctx
        )
        doc = desktop.loadComponentFromURL(
            "file://" + csv_path,
            "_blank",
            0,
            (
                pv("Hidden", True),
                pv("FilterName", "Text - txt - csv (StarCalc)"),
                pv("FilterOptions", "44,34,76,1"),
            ),
        )
        doc.storeToURL(
            "file://" + os.path.join(out, "xlsx-password-protected.xlsx"),
            (
                pv("FilterName", "Calc MS Excel 2007 XML"),
                pv("Password", "geheim"),
                pv("EncryptFile", True),
            ),
        )
        doc.close(False)
        desktop.terminate()
    finally:
        soffice.wait(timeout=30)
        shutil.rmtree(work, ignore_errors=True)


if __name__ == "__main__":
    mode, out = sys.argv[1], sys.argv[2]
    {"parquet": parquet, "protected-xlsx": protected_xlsx}[mode](out)
    print(f"generate.py {mode}: done")

File: tests/fixtures/generate.mjs
// The fixture generator. Run: `node tests/fixtures/generate.mjs`
//
// WHY THESE FILES ARE COMMITTED RATHER THAN BUILT AT TEST TIME. Every other
// fixture in this repo is written by `write-excel-file` or `hyparquet-writer`
// inside the test that reads it, which is the better arrangement when it is
// possible: the shape under test is stated in the test itself. These five are
// the ones neither writer can produce at all. Committing the bytes *and* this
// script answers the only real objection to a checked-in binary — that nobody
// can see what is inside it or make another one.
//
// WHAT EACH FILE IS, AND WHAT MADE IT.
//
//   parquet-int96-timestamp.parquet              pyarrow 25.0.0
//     One INT96 timestamp column (the legacy Impala/older-Spark encoding) and
//     one UTF8 column. pyarrow writes INT96 only behind
//     `use_deprecated_int96_timestamps=True`; no JavaScript writer emits it.
//
//   parquet-decimal-fixed-len-byte-array.parquet pyarrow 25.0.0
//     `decimal128(38, 2)`. The physical backing of a DECIMAL follows its
//     precision — <= 9 is INT32, <= 18 INT64, beyond that FIXED_LEN_BYTE_ARRAY —
//     and `hyparquet-writer` supports only the two integer backings.
//
//   parquet-duplicate-column-name.parquet        pyarrow 25.0.0
//     Two top-level columns both named `Betrag`, one INT64 and one UTF8. Legal
//     Parquet, and unwritable with `hyparquet-writer`, which resolves its
//     `columnData` against the schema by name. The differing physical types are
//     deliberate: a reader that confuses the two is caught by the values.
//
//   xlsx-password-protected.xlsx                 LibreOffice 25.x via UNO
//     A real OOXML-encrypted workbook, password `geheim`. Not a zip at all — an
//     encrypted workbook is a CFB container starting `D0 CF 11 E0`. Verified
//     encrypted and decryptable with `msoffcrypto-tool` when it was made.
//     LibreOffice's `--convert-to` filter options silently ignore a password for
//     xlsx; the UNO media descriptor is the route that works.
//
//   xlsx-zero-sheets.xlsx                        this script, via fflate
//     A workbook whose `xl/workbook.xml` declares `<sheets/>` — no sheet at all.
//     Built by unzipping an ordinary workbook, rewriting that one element and
//     rezipping. Excel will not produce one; it exists to exercise the reader's
//     empty-workbook branch, and what it actually proves is recorded in the
//     test that reads it.
//
// PREREQUISITES, and they are deliberately not project dependencies:
//   * pyarrow, in a throwaway venv — `pip install` is PEP-668-blocked system
//     wide on this machine, so: `python3 -m venv /tmp/qb && /tmp/qb/bin/pip
//     install pyarrow`, then point PYARROW_PYTHON at /tmp/qb/bin/python.
//   * a system `python3` carrying the `uno` module (ships with LibreOffice) and
//     `soffice` on PATH.
//   * `fflate`, used here only, and reached through `read-excel-file`'s own
//     dependency rather than added to package.json. If that ever stops being
//     true this script needs its own install line; nothing shipped depends on it.

import { execFileSync } from 'node:child_process'
import { existsSync, readFileSync, writeFileSync } from 'node:fs'
import { dirname, join } from 'node:path'
import { fileURLToPath } from 'node:url'
import { strFromU8, strToU8, unzipSync, zipSync } from 'fflate'
import writeXlsxFile from 'write-excel-file/node'

const OUT = dirname(fileURLToPath(import.meta.url))
const PYARROW_PYTHON = process.env.PYARROW_PYTHON ?? '/tmp/qb/bin/python'
const SYSTEM_PYTHON = process.env.SYSTEM_PYTHON ?? 'python3'

const run = (what, interpreter, mode) => {
  try {
    execFileSync(interpreter, [join(OUT, 'generate.py'), mode, OUT], { stdio: 'inherit' })
  } catch (failure) {
    console.error(`\n${what} failed (${interpreter}). See the prerequisites at the top.\n`, failure.message)
    process.exitCode = 1
  }
}

run('Parquet fixtures', PYARROW_PYTHON, 'parquet')
run('Password-protected workbook', SYSTEM_PYTHON, 'protected-xlsx')

// The zero-sheet workbook. Written here rather than in Python because the whole
// operation is three lines against a zip, and because it is the one fixture that
// starts from a workbook this repo's own writer produced.
{
  const writer = await writeXlsxFile(
    [[{ value: 'Kunde', type: String }], [{ value: 'Anna', type: String }]],
    { sheet: 'Umsatz' },
  )
  const buffer = await writer.toBuffer()
  const entries = unzipSync(new Uint8Array(buffer))
  const workbook = strFromU8(entries['xl/workbook.xml'])
  if (!/<sheets>/.test(workbook)) throw new Error('no <sheets> element to strip — writer changed')
  entries['xl/workbook.xml'] = strToU8(workbook.replace(/<sheets>.*?<\/sheets>/s, '<sheets/>'))
  writeFileSync(join(OUT, 'xlsx-zero-sheets.xlsx'), zipSync(entries))
  console.log('generate.mjs: xlsx-zero-sheets.xlsx done')
}

for (const name of [
  'parquet-int96-timestamp.parquet',
  'parquet-decimal-fixed-len-byte-array.parquet',
  'parquet-duplicate-column-name.parquet',
  'xlsx-password-protected.xlsx',
  'xlsx-zero-sheets.xlsx',
]) {
  const path = join(OUT, name)
  const state = existsSync(path) ? `${readFileSync(path).byteLength} bytes` : 'MISSING'
  console.log(`  ${name.padEnd(46)} ${state}`)
}
